builder/dockerfile/dispatchers.go

```diff
@@ -235,24 +250,69 @@ func (b *Builder) getFromImage(dispatchState *dispatchState, shlex *ShellLex, na
 		return nil, err
 	}
 
-	if im, ok := b.imageContexts.byName[name]; ok {
-		if len(im.ImageID()) > 0 {
-			return im, nil
-		}
-		// FROM scratch does not have an ImageID
-		return nil, nil
+	if im, ok := b.buildStages.getByName(name); ok {
+		return im, nil
 	}
 
 	// Windows cannot support a container with no base image.
 	if name == api.NoBaseImageSpecifier {
 		if runtime.GOOS == "windows" {
 			return nil, errors.New("Windows does not support FROM scratch")
 		}
-		dispatchState.imageID = ""
-		dispatchState.noBaseImage = true
-		return nil, nil
+		return scratchImage, nil
+	}
+	imageMount, err := b.imageMounter.Get(name)
```

Comment:
That's true, but the purpose of the struct is still to manage the unmounts required because of the API we are using. I don't know what else to call it